### DISCUSSION TITLE
Parallel consumer with spring boot

### DISCUSSION THREAD

**Discussion by amber13574** [confluentinc/parallel-consumer#671] [Q&A]
Hi Folks,

I'm currently working on my service where I am testing how the parallel-consumer library (https://github.com/confluentinc/parallel-consumer) behaves with spring. However, I've encountered an issue while setting up the factory container customizer. I want to set my message listener, which is of type ParallelStreamProcessor. However, the method setupMessageListener only expects AcknowledgingMessageListener or MessageListener. Is there any way to address this? One option is to create a customized listener by extending one of the existing listeners, but I'm open to other solutions.

```
/**
     * Set the message listener; must be a {@link org.springframework.kafka.listener.MessageListener}
     * or {@link org.springframework.kafka.listener.AcknowledgingMessageListener}.
     * @param messageListener the listener.
     */
    public void setMessageListener(Object messageListener) {
        this.messageListener = messageListener;
        adviseListenerIfNeeded();
    }
```
This is how I am creating beans

```
@Bean
  public ParallelStreamProcessor<String, String> parallelStreamProcessor(Consumer<String, String> kafkaConsumer) {

    ParallelConsumerOptions<String, String> options = ParallelConsumerOptions.<String, String>builder()
        .ordering(ParallelConsumerOptions.ProcessingOrder.KEY)
        .maxConcurrency(16)
        .commitMode(ParallelConsumerOptions.CommitMode.PERIODIC_CONSUMER_SYNC)
        .consumer(kafkaConsumer)
        .build();

    return ParallelStreamProcessor.createEosStreamProcessor(options);
  }
```

```
  @Bean
  public ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory(
      ConsumerFactory<String, String> consumerFactory,
      ParallelStreamProcessor<String, String> parallelStreamProcessor) {
    ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
    factory.setConsumerFactory(consumerFactory);

    factory.setContainerCustomizer(container -> {
      container.setupMessageListener(parallelStreamProcessor);
      container.setAutoStartup(true);
    });

    return factory;
  }
```
I expect alternate solution to resolve this issue.

As I can see there are no example present at the moment , but feel free to add example If anyone has implemented the confluent  parallel-consumer with spring.

**Comment by rkolesnev** [comment]
Hi @amber13574,
Honestly i haven't had integrated Parallel Consumer with Spring through Spring factory / customizers properly before.
Have you figured out a good way to do it? 
I would really appreciate a full example to add it to docs and integration examples.

**Comment by astubbs** [comment]
@amber13574 - the container customizer is the wrong seam, and that is why it will not take a `ParallelStreamProcessor`: PC *is* the container. It owns the poll loop, the worker threads and the offset commits, so it replaces Spring Kafka's listener container rather than plugging into it.

The shape that works: keep Spring's `ConsumerFactory` for building the `KafkaConsumer` (with `enable.auto.commit=false`, which PC requires), build the processor from it as you already do, and give the processor a lifecycle bean - `SmartLifecycle` or `@PostConstruct`/`@PreDestroy` - whose start calls `subscribe(...)` then `poll(...)` with your function, and whose stop calls `closeDrainFirst()`. Do not register it as a message listener at all. For Actuator, this tracker's #403 has a `HealthIndicator` over `isClosedOrFailed()` that people run in production.

Where the maintained fork stands: nothing about this changed on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0), except that MDC values set before `poll()` now reach the worker threads, which matters for Spring's tracing. There is no Spring Boot example in the repository, and @rkolesnev's request for one still stands - it is a documentation gap the fork will close, and a contributed one from a working setup would be taken gladly. A health-check surface, an open pull request on the fork and not yet merged (astubbs/parallel-consumer#226), adds `getHealth()` to the interface so the indicator no longer needs a cast to the concrete class.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
